crates/config/src/endpoints.rs

```diff
@@ -134,6 +232,86 @@ impl<'de> Deserialize<'de> for RpcEndpoint {
     }
 }
 
+impl Into<RpcEndpointType> for RpcEndpoint {
+    fn into(self) -> RpcEndpointType {
+        RpcEndpointType::String(self)
+    }
+}
+
+/// Rpc endpoint configuration variant
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct RpcEndpointConfig {
+    /// endpoint url or env
+    pub endpoint: RpcEndpoint,
+
+    /// The number of retries.
+    pub retries: Option<u32>,
+
+    /// Initial retry backoff.
+    pub retry_backoff: Option<u64>,
+
+    /// The available compute units per second.
+    ///
+    /// See also <https://docs.alchemy.com/reference/compute-units#what-are-cups-compute-units-per-second>
+    pub compute_units_per_second: Option<u64>,
+}
+
+impl fmt::Display for RpcEndpointConfig {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let RpcEndpointConfig { endpoint, retries, retry_backoff, compute_units_per_second } = self;
+
+        write!(f, "{}", endpoint)?;
+
+        if let Some(retries) = retries {
+            write!(f, ", retries={}", retries)?;
+        }
+
+        if let Some(retry_backoff) = retry_backoff {
+            write!(f, ", retry_backoff={}", retry_backoff)?;
+        }
+
+        if let Some(compute_units_per_second) = compute_units_per_second {
+            write!(f, ", compute_units_per_second={}", compute_units_per_second)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl Serialize for RpcEndpointConfig {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut map = serializer.serialize_map(Some(4))?;
+
+        map.serialize_entry("endpoint", &self.endpoint)?;
+        map.serialize_entry("retries", &self.retries)?;
+        map.serialize_entry("retry_backoff", &self.retry_backoff)?;
+        map.serialize_entry("compute_units_per_second", &self.compute_units_per_second)?;
+
+        map.end()
```

Comment:
here we can also use the enum, and use RpcEndpoint directly if all the other fields are None

Reply:
could you please give an example of how to write this?